predicates/forwarded/forwarded.go

```diff
@@ -0,0 +1,156 @@
+
+Examples:
+
+    // only match requests to "example.com"
+    example1: ForwardedHost("example.com") -> "http://example.org";
+
+    // only match requests to http
+    example2: ForwardedProtocol("http") -> "http://example.org";
+
+    // only match requests to https
+    example3: ForwardedProtocol("https") -> "http://example.org";
+*/
+package forwarded
+
+import (
+	"github.com/zalando/skipper/predicates"
+	"github.com/zalando/skipper/routing"
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+const (
+	NameHost  = "ForwardedHost"
+	NameProto = "ForwardedProtocol"
+)
+
+const (
+	HTTP  string = "http"
+	HTTPS string = "https"
+)
+
+type hostPredicateSpec struct{}
+
+type protoPredicateSpec struct{}
+
+type hostPredicate struct {
+	host *regexp.Regexp
+}
+
+func (p *hostPredicateSpec) Create(args []interface{}) (routing.Predicate, error) {
+
+	if len(args) != 1 {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	value, ok := args[0].(string)
+	if !ok {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	if value == "" {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	re, err := regexp.Compile(value)
+	if err != nil {
+		return nil, err
+	}
+
+	return hostPredicate{host: re}, err
+}
+
+type protoPredicate struct {
+	proto string
+}
+
+func (p *protoPredicateSpec) Create(args []interface{}) (routing.Predicate, error) {
+
+	if len(args) != 1 {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	value, ok := args[0].(string)
+	if !ok {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	value = strings.ToLower(value)
+
+	switch value {
+	case HTTP, HTTPS:
+	default:
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	return protoPredicate{proto: value}, nil
+}
+
+func NewForwardedHost() routing.PredicateSpec  { return &hostPredicateSpec{} }
+func NewForwardedProto() routing.PredicateSpec { return &protoPredicateSpec{} }
+
+func (p *hostPredicateSpec) Name() string {
+	return NameHost
+}
+
+func (p *protoPredicateSpec) Name() string {
+	return NameProto
+}
+
+func (p hostPredicate) Match(r *http.Request) bool {
+
+	fh := r.Header.Get("Forwarded")
+
+	if fh == "" {
+		return false
+	}
+
+	fw := parseForwarded(fh)
+
+	return p.host.MatchString(fw.host)
+}
+
+func (p protoPredicate) Match(r *http.Request) bool {
+
+	fh := r.Header.Get("Forwarded")
+
+	if fh == "" {
+		return false
+	}
+
+	fw := parseForwarded(fh)
+
+	return p.proto == fw.proto
+}
+
+type forwarded struct {
+	host  string
+	proto string
+}
+
+func parseForwarded(fh string) *forwarded {
+
+	f := &forwarded{}
+
+	for _, forwardedPair := range strings.Split(fh, ";") {
+		if tv := strings.SplitN(forwardedPair, "=", 2); len(tv) == 2 {
+			token, value := tv[0], tv[1]
+			token = strings.TrimSpace(token)
+			value = strings.TrimSpace(strings.Trim(value, `"`))
+			switch strings.ToLower(token) {
+			case "proto":
+				f.proto = value
+			case "host":
+				f.host = value
+			}
+		}
+	}
```

Comment:
An alternative implementation could be to transform predicate argument into regexp and match the header value against it. Since regexp is used anyway we can avoid parsing the structure altogether.

Reply:
I was trying to think about this.... but building everythign into a single regexp may be actually harder:
- How do you combine the hole thing into a single regexp if what you get in the host is something like this?

`(^|www\.|subdomain)\.example\.org$`

The end regexp should not have ^ or $, but `host=` and `($|;)` instead, and that string manipulation could get tricky.

I'll leave it like it is now, we can always change the implementation given that all tests keep passing.

Reply:
Agree, rewriting regexp does not seem right